src/core/nixl_agent.cpp

```diff
@@ -114,23 +112,26 @@ nixlAgent::nixlAgent(const std::string &name,
 
     if (data->useEtcd || cfg.useListenThread) {
         data->commThreadStop = false;
-        data->commThread = std::thread(&nixlAgentData::commWorker, data, this);
+        data->commThread =
+            std::thread(&nixlAgentData::commWorker, data.get(), this);
     }
 }
 
 nixlAgent::~nixlAgent() {
-    if (data->useEtcd || data->config.useListenThread) {
+    if (data && (data->useEtcd || data->config.useListenThread)) {
         data->commThreadStop = true;
         if(data->commThread.joinable()) data->commThread.join();
 
         if(data->config.useListenThread) {
             if(data->listener) delete data->listener;
```

Comment:
Having to manually deallocate parts of 'data' on cleanup kinda defeats the purpose of automatic memory management with RAII. Looking at the code it is not immediately obvious to me why data->listener even needs to be a pointer in the first place, so consider refactoring it to be by-value member of nixlAgentData.

Reply:
Agree that nixlAgentData also needs refactoring. However IMO we need to protect nixlAgent usage first, since refactoring nixlAgentData can be a deeper change.

Reply:
I suggested this because making listener either by-value member, or, if not possible, another unique_ptr seemed like a trivial change to me that really makes nixlAgentData compatible with RAII idiom, but I'm not insisting on it.